### DISCUSSION TITLE
Could arguments in shortcuts allow brackets?

### DISCUSSION THREAD

**Discussion by GiovanniSalmeri** [datenstrom/community#976] [Ask a question] [answered]
Currently the regex's at markdown.php:3879-3880 do not allow brackets (more exactly: closing brackets) in the arguments of shortcuts. This is hardly relevant for most extensions, but is a serious drawback for [Math](https://github.com/GiovanniSalmeri/yellow-math/), since brackets are an important part of mathematical syntax. After different experiments, I thought of the following solution: brackets could be allowed in an argument if enclosed in quotes. This would require only different regex's:

```php
$text = preg_replace_callback("/^\s*\[(\w+)((?:\s+(?:[^\]\"\s]+|\"(?:[^\"]|\"\"|\\\\\")*\"))*\s*)\]\s*$/", array($this, "_doAutoLinks_shortcutBlock_callback"), $text);
$text = preg_replace_callback("/\[(\w+)((?:\s+(?:[^\]\"\s]+|\"(?:[^\"]|\"\"|\\\\\")*\"))*\s*)\]/", array($this, "_doAutoLinks_shortcutInline_callback"), $text);
```

The regex, even if it seems complicated, is actually straigthforward. It states that after the name of the extensions `(\w+)` the arguments can be either a sequence of characters different from closing bracket, quote or space (`[^\]\"\s]+`), or a sequence of characters enclosed in quotes: either different from quote (`[^\"]`), or escaped quotes (`\"\"`), or backslash-escaped quotes (`\\\\\"`). I followed the definition of CSV in [RFC4180](https://datatracker.ietf.org/doc/html/rfc4180) *and* the extension in [PHP `str_getcsv`](https://www.php.net/manual/en/function.str-getcsv.php).

Unless I miss something (very possible 😁), this solution would negatively affect in no way any extension, since the parsing of the arguments (by means of `str_getcsv`) would be left untouched in core.php. Only a new character (`]`) would be allowed in arguments enclosed in quotes. For example one could write: `[image photo.jpg "Beautiful photo [by Alejandro Escamilla]" rounded]`. But the great advantage would be for Math, in its README I could simply write: "Wrap the whole expression in quotes if there are spaces, brackets, or quotes".

Actually this is my fourth attempt at solving this issue:

1. Firstly I defined a symbol `rbracket` inside Math: but actually it didn't work, because it was interpreted too late
2. Then (in the present version of Math) I devised an escape `%|`: it works, but is ugly and difficult to remember
3. Later I tried with an escaped `\]` (unpublished): but it required *two* modifications in markdown.php: a rather complex regex and later a substitution in `_doAutoLinks_shortcutBlock_callback` and `_doAutoLinks_shortcutInline_callback`
4. The fourth attempt is what I have described above.

Sorry for the long message! What do you think about that?

**Comment by markseuffert** [comment, marked as answer]
No, it's better to use URL encoding or character entity.
